### ISSUE TITLE
package instead of project name in toml file

### ISSUE THREAD

**Issue by guiwitz** [tlambert03/pyrepo-cookiecutter#13]
Trying to install packages made with the cookiecutter in editable mode, I noticed that for cases where the ```project_name``` and ```project_slug``` are different, packages can then not be imported. I think the problem is here: https://github.com/tlambert03/pyrepo-cookiecutter/blob/700a4fd096dd3dfb2e6c3a4d767b85a1f453a076/%7B%7Bcookiecutter.project_name%7D%7D/pyproject.toml#LL8C5-L8C5 

In the [official documentation](https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata) the project name (main folder) can be totally different from the package name (i.e. ```slug_name```). But in the ```pyproject.toml``` [file](https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata), the ```name``` should be the package name not the project name. It's probably unusual to have different names (and maybe it works with dashes and underscores) but I guess this should be fixed in the ```pyproject.toml``` file of the cookiecutter? At least when I fix the ```name``` in the toml file, the package becomes findable.

**Comment by tlambert03** [comment]
> But in the pyproject.toml [file](https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata), the name should be the package name not the project name.

where are you getting that?  The linked page on python metadata says:

> `name` is the distribution name of your package. This can be any name as long as it only contains letters, numbers, ., _ , and -. It also must not already be taken on PyPI. Be sure to update this with your username for this tutorial, as this ensures you won’t try to upload a package with the same name as one which already exists.

"distribution name" here means the project name, not the slug... 

that said, I'm happy to try to fix this!  Will I be able to reproduce it just be using something with a dash in the name?  or do they need to be rather different?

**Comment by guiwitz** [comment]
I agree it's confusing, and I'm going to assume that it's just the tutorial that is badly formulated. If you check the beginning of the tutorial, they explain the structure of a project (https://packaging.python.org/en/latest/tutorials/packaging-projects/#a-simple-project). As you see below, what they call the distribution name, ```example_package_YOUR_USERNAME_HERE```, is the first folder in ```src``` which is the ```slug_name``` in the cookiecutter.
<img width="822" alt="Screenshot 2023-06-07 at 23 17 50" src="https://github.com/tlambert03/pyrepo-cookiecutter/assets/4622767/067ad7e1-68ea-4d7d-9389-56eb4ba6861b">
<img width="819" alt="Screenshot 2023-06-07 at 23 23 47" src="https://github.com/tlambert03/pyrepo-cookiecutter/assets/4622767/39124400-aa74-4ce8-8f5c-31c2b8456246">

The problem with this is that it makes it impossible to have different names for project and package like ```scikit-image``` and ```skimage```, so it must be wrong. However the issue with the editable install remains. In the current situation, if you use rather different names (not just a dash) like in the tutorial, the package seems installed (with the project name, i.e. ```packaging_tutorial```) but it can't be imported using the slug name.

**Comment by tlambert03** [comment]
Ok thanks, lemme look into it :)

**Comment by tlambert03** [comment]
ok, the trick is to uncomment these lines:

https://github.com/tlambert03/pyrepo-cookiecutter/blob/700a4fd096dd3dfb2e6c3a4d767b85a1f453a076/%7B%7Bcookiecutter.project_name%7D%7D/pyproject.toml#L67-L69

without those lines, hatch looks in the root for packages.  so when you build a wheel for project `some-project` with module `src/my_module` you get this:

```
$ python -m build --wheel
* Creating venv isolated environment...
* Installing packages in isolated environment... (hatch-vcs, hatchling)
* Getting build dependencies for wheel...
* Building wheel...
Successfully built some_project-0.1.dev1+ge38455e-py3-none-any.whl

$ unzip dist/some_project-* -d wheel
Archive:  dist/some_project-0.1.dev1+ge38455e-py3-none-any.whl
  inflating: wheel/setup.py
  inflating: wheel/src/my_module/__init__.py
  inflating: wheel/wheel/setup.py
  inflating: wheel/wheel/src/my_module/__init__.py
  inflating: wheel/some_project-0.1.dev1+ge38455e.dist-info/METADATA
  inflating: wheel/some_project-0.1.dev1+ge38455e.dist-info/WHEEL
  inflating: wheel/some_project-0.1.dev1+ge38455e.dist-info/licenses/LICENSE
  inflating: wheel/some_project-0.1.dev1+ge38455e.dist-info/RECORD

$ tree wheel
wheel
├── setup.py
├── some_project-0.1.dev1+ge38455e.dist-info
│   ├── METADATA
│   ├── RECORD
│   ├── WHEEL
│   └── licenses
│       └── LICENSE
├── src
│   └── my_module
│       └── __init__.py
└── wheel
    ├── setup.py
    └── src
        └── my_module
            └── __init__.py

8 directories, 8 files

$ cat wheel/some_project-*/RECORD
setup.py,sha256=FD1GHc9zioaZr7D5s4am6BIe9bTDDkkTlFSpAHO5HrI,583
src/my_module/__init__.py,sha256=rjlqfdSu7tFINPqn8MjJv0VqpKZklQIghOFlBazpfRU,255
wheel/setup.py,sha256=FD1GHc9zioaZr7D5s4am6BIe9bTDDkkTlFSpAHO5HrI,583
wheel/src/my_module/__init__.py,sha256=rjlqfdSu7tFINPqn8MjJv0VqpKZklQIghOFlBazpfRU,255
some_project-0.1.dev1+ge38455e.dist-info/METADATA,sha256=rbmO2h9bf-7vzm7GFsCQZh-6P-riMYqECkFjQcwsGOc,1945
some_project-0.1.dev1+ge38455e.dist-info/WHEEL,sha256=KGYbc1zXlYddvwxnNty23BeaKzh7YuoSIvIMO4jEhvw,87
some_project-0.1.dev1+ge38455e.dist-info/licenses/LICENSE,sha256=strNG-vpfkIeqjlPMhD6OR4AGiA7-5X9fDOQprfa_E0,1513
some_project-0.1.dev1+ge38455e.dist-info/RECORD,,
```

by contrast, when you tell hatch where the src directory is, you get this (note the top level `my_module` in the RECORD as well as the tree):

```
$ python -m build --wheel
* Creating venv isolated environment...
* Installing packages in isolated environment... (hatch-vcs, hatchling)
* Getting build dependencies for wheel...
* Building wheel...
Successfully built some_project-0.1.dev1+ge38455e.d20230608-py3-none-any.whl

$ unzip dist/some_project-* -d wheel
Archive:  dist/some_project-0.1.dev1+ge38455e.d20230608-py3-none-any.whl
  inflating: wheel/my_module/__init__.py
  inflating: wheel/my_module/py.typed
  inflating: wheel/some_project-0.1.dev1+ge38455e.d20230608.dist-info/METADATA
  inflating: wheel/some_project-0.1.dev1+ge38455e.d20230608.dist-info/WHEEL
  inflating: wheel/some_project-0.1.dev1+ge38455e.d20230608.dist-info/licenses/LICENSE
  inflating: wheel/some_project-0.1.dev1+ge38455e.d20230608.dist-info/RECORD

$ tree wheel
wheel
├── my_module
│   ├── __init__.py
│   └── py.typed
└── some_project-0.1.dev1+ge38455e.d20230608.dist-info
    ├── METADATA
    ├── RECORD
    ├── WHEEL
    └── licenses
        └── LICENSE

4 directories, 6 files

$ /bin/cat wheel/some_project-*/RECORD
my_module/__init__.py,sha256=rjlqfdSu7tFINPqn8MjJv0VqpKZklQIghOFlBazpfRU,255
my_module/py.typed,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
some_project-0.1.dev1+ge38455e.d20230608.dist-info/METADATA,sha256=X9TlVimP4fIC-kzNtNS7x9O36zS-mqgt-lM6z3HYCBc,1955
some_project-0.1.dev1+ge38455e.d20230608.dist-info/WHEEL,sha256=KGYbc1zXlYddvwxnNty23BeaKzh7YuoSIvIMO4jEhvw,87
some_project-0.1.dev1+ge38455e.d20230608.dist-info/licenses/LICENSE,sha256=strNG-vpfkIeqjlPMhD6OR4AGiA7-5X9fDOQprfa_E0,1513
some_project-0.1.dev1+ge38455e.d20230608.dist-info/RECORD,,
```

----

so, to summarize, I still think that the `project.name` field in pyproject is meant to be the package as you would have it on PyPI... and _not_ your top level module.  It's possible for a single package to offer multiple top level modules, so that also would be hard to do with `project.name` being the import path.    I'll update the cookiecutter to uncomment those lines

**Comment by tlambert03** [comment]
[06e282b](https://github.com/tlambert03/pyrepo-cookiecutter/commit/06e282b86a6637941fd5c46827383561408cd26f) applies this to the simple branch as well

**Comment by guiwitz** [comment]
Great, thanks for looking into this so quickly! I'm glad that everything makes sense again!
